source/common/router/vhds.cc

```diff
@@ -36,14 +37,23 @@ VhdsSubscription::VhdsSubscription(
           fmt::format("VhdsConfigSubscription {}", config_update_info_->routeConfigName()),
           [this]() { subscription_->start({config_update_info_->routeConfigName()}, true); }),
       route_config_providers_(route_config_providers) {
-  const auto& config_source = config_update_info_->routeConfiguration()
-                                  .vhds()
-                                  .config_source()
-                                  .api_config_source()
-                                  .api_type();
-  if (config_source != envoy::config::core::v3::ApiConfigSource::DELTA_GRPC) {
-    throw EnvoyException("vhds: only 'DELTA_GRPC' is supported as an api_type.");
+
+  if (config_update_info_->routeConfiguration().vhds().config_source().has_ads()) {
+    auto adsMux = factory_context.clusterManager().adsMux().get();
+    if (typeid(*adsMux) != typeid(Envoy::Config::NewGrpcMuxImpl)) {
```

Comment:
Relying on RTTI here is fairly fragile. I think `subscriptionFromConfigSource` should take a flag indicating if a "on-demand" capability is required and have this resolved there.

Reply:
Yes, I hesitated about using RTTI here too. I seems that moving this logic to subscriptionFromConfigSource is much better~

Reply:
I have a question about subscriptionFromConfigSource. I saw in subscriptionFromConfigSource,It is easy to get whether api_config_source.api_type is delta. but if we want to support ads, we should check whether the ads is delta or not, but I can not find the information in subscriptionFromConfigSource, so do not find a way to determine whether the ads is delta. what about add a interface in GrpcMux to indicate whether it is delta or not?  instead of using RTTI?

Reply:
Yeah, I agree that fundamentally this needs to live on the GrpcMux interface. Delta though is an implementation detail, the real question is does this support namespace matching (see addWatch()). @dmitri-d is there any reason that VHDS wouldn't "just work" with SotW when we converge NewGrpcMuxImpl and GrpcMuxImpl?

Reply:
> is there any reason that VHDS wouldn't "just work" with SotW when we converge NewGrpcMuxImpl and GrpcMuxImpl?

Nothing immediately obvious. I'm not sure I understand this correctly though -- why would VHDS over SotW xDS be needed? RDS config is (almost) always up-to date when SotW xDS is used, why would VHDS be needed in that case?

Reply:
It's not really whether it's needed, just wondering if there's a reason that regular GrpcMuxImpl wouldn't be able to implement it. This is another case of "two GrpcMuxImpls bad!".

Reply:
> > is there any reason that VHDS wouldn't "just work" with SotW when we converge NewGrpcMuxImpl and GrpcMuxImpl?
> 
> Nothing immediately obvious. I'm not sure I understand this correctly though -- why would VHDS over SotW xDS be needed? RDS config is (almost) always up-to date when SotW xDS is used, why would VHDS be needed in that case?

hi, @dmitri-d , this PR is make VHDS can use Ads (delta api too).  Not over SotW xDS or Ads. otherwise the VHDS will create another stream.

Reply:
the ads config is like this:  
ads_config:
    api_type: DELTA_GRPC
    transport_api_version: V3
    grpc_services:
    - envoy_grpc:
        cluster_name: xds_cluster

So do  you have any concerns about using VHDS in the ADS stream?  @dmitri-d

Reply:
> This is another case of "two GrpcMuxImpls bad!".

I'm on it, was away on a holiday. 

> this PR is make VHDS can use Ads (delta api too). Not over SotW xDS or Ads. otherwise the VHDS will create another stream.

It's not possible to run VHDS over ADS, regardless whether it uses delta- or SotW xDS. There is no way to associate a given VHDS request with an RDS config when multiple VHDS subscriptions are multiplexed over an ADS connection. B/c of this limitation, I don't think we should be enabling VHDS over ADS.

Reply:
Hmm, actually, I think requests are fine, aliases contain prefix/namespace which is the name of the route config. The initial request/response exchange in a newly started VHDS subscription does have an issue: 
 - the initial VHDS request is empty (as in has no aliases/resources names to subscribe to), so the server won't be able to associate the subscription with an RDS config at this point.
 - The initial server response is also empty; if there are several VHDS subscriptions starting, how do we select which one should accept the response that just arrived?

Reply:
Do you mean the response from the XDS management server? the aliases is make up of  route_config_name + "/" + domain . So I seems that we can use  route_config_name to select VHDS subscriptions ?

Reply:
This is essentially the problem solved by collections in the https://github.com/envoyproxy/envoy/issues/11264 proposal. I wonder if the semantics in the absence of VHDS collections would be that we only have a single VHDS subscription per GrpcMux, and all subscriptions just watch on that?

Reply:
> Hmm, actually, I think requests are fine once VHDS subscription is started -- aliases contain prefix/namespace which is the name of the route config. The initial request/response exchange in a newly started VHDS subscription does have an issue:
> 
> * the initial VHDS request is empty (as in has no aliases/resources names to subscribe to), so the server won't be able to associate the subscription with an RDS config at this point.
> * The initial server response is also empty; if there are several VHDS subscriptions starting, how do we select which one should accept the response that just arrived?

sorry I misunderstood what you said.  Yes the  initial request is empty and the response is empty too. so why  we need to accept these empty response? It seems useless.  the following on-demand request contain  route_config_name so It is easy to distinguish

Reply:
> This is essentially the problem solved by collections in the #11264 proposal. I wonder if the semantics in the absence of VHDS collections would be that we only have a single VHDS subscription per GrpcMux, and all subscriptions just watch on that?

That probably depends on whether collection state is a thing and at what point we want it to get pushed from the server. We could handle it lazily (ie return the state in the response to the first query), but there might be performance implications and corner cases here (I can think of one already).

Reply:
This seems a valid point, can we (or do we already) ignore empty responses? A test along these lines would be the ultimate evidence.

Reply:
I check the code，empty  responses will be ignored already.
this is because empty responses will trigger  wildcard_watches_  updates. see below:

`
  if (added_resources.empty() && removed_resources.empty()) {
    for (auto& cur_watch : wildcard_watches_) {
      cur_watch->callbacks_.onConfigUpdate({}, {}, system_version_info);
    }
  }
`
in fact， vhds watches use route_config_name， so the wildcard_watches_ will be empty.  the empty  responses will be ignore.

Reply:
> This seems a valid point, can we (or do we already) ignore empty responses? A test along these lines would be the ultimate evidence.

An empty response should be a nop and @Agent-Tao reply above confirms that (although I'd need to check how acks are handled in this case). This would obviously affect server implementation and need to be documented. Ultimately, I agree with @htuch: a test demonstrating the behaviour is needed.

Reply:
So It seems that VHDS could use  delta ADS?  if so, I will try do  remove  RTTI, mayby adding a Method in GrpcMux to indicate is delta or not,  document the server implementation about this change, and also some test about this behaviour? @htuch @dmitri-d

Reply:
SGTM

Reply:
sgtm

Reply:
this pr can by review now, thanks @htuch @dmitri-d